src/backend/distributed/connection/connection_management.c

```diff
@@ -527,19 +530,188 @@ FindAvailableConnection(dlist_head *connections, uint32 flags)
+				 transaction->transactionFailed &&
+				 (transaction->transactionCritical ||
+				  !dlist_is_empty(&connection->referencedPlacements)))
+		{
+			/*
+			 * If a connection is executing a critical transaction or accessed any
+			 * placements, we should not continue the execution.
+			 */
+			ReportConnectionError(connection, ERROR);
+		}
+		else
+		{
+			/*
+			 * If a connection is not usable and we do not need to throw an error,
+			 * close at the end of the transaction.
+			 */
+			connection->forceCloseAtTransactionEnd = true;
+		}
+	}
+
+	return healthyConnections;
+}
+
+
+/*
+ * RemoteSocketClosed returns true if the remote socket is closed.
+ *
+ * Note that we rely on WL_SOCKET_CLOSED, which is only available
+ * for PG 15+ and for kernels that support the feature.
+ */
+static bool
+RemoteSocketClosed(MultiConnection *connection)
+{
+	bool socketClosed = false;
+#if PG_VERSION_NUM >= PG_VERSION_15
+
+	if (!WaitEventSetCanReportClosed())
+	{
+		/* the kernel build does not support WL_SOCKET_CLOSED */
+		return false;
+	}
+
+	int eventCount = 0;
+	WaitEvent events[3];
+
+	WaitEventSet *waitEventSet =
+		CreateWaitEventSet(CurrentMemoryContext, 3);
+	int sock = PQsocket(connection->pgConn);
+	int waitEventSetIndex =
+		CitusAddWaitEventSetToSet(waitEventSet, WL_SOCKET_CLOSED, sock,
+								  NULL, (void *) connection);
+
+	if (waitEventSetIndex == WAIT_EVENT_SET_INDEX_FAILED)
+	{
+		/*
+		 * Inform failed to add to wait event set with a debug message as this
+		 * is too detailed information for users. We let the code flow just in
+		 * case it can successfully finish the execution.
+		 */
+		ereport(DEBUG1, (errcode(ERRCODE_CONNECTION_FAILURE),
+						 errmsg("Adding wait event for node %s:%d failed. "
+								"The socket was: %d",
+								connection->hostname,
+								connection->port, sock)));
+
+		FreeWaitEventSet(waitEventSet);
+
+		return false;
+	}
+
+	CitusAddWaitEventSetToSet(waitEventSet, WL_POSTMASTER_DEATH, PGINVALID_SOCKET, NULL,
+							  NULL);
+	CitusAddWaitEventSetToSet(waitEventSet, WL_LATCH_SET, PGINVALID_SOCKET, MyLatch,
+							  NULL);
+
+	long timeout = 0; /* do not wait at all */
+	int eventIndex = 0;
+
+retry:
+	eventCount =
+		WaitEventSetWait(waitEventSet, timeout, events,
+						 3, WAIT_EVENT_CLIENT_READ);
+	for (; eventIndex < eventCount; eventIndex++)
+	{
+		WaitEvent *event = &events[eventIndex];
+
+		if (event->events & WL_POSTMASTER_DEATH)
+		{
+			ereport(ERROR, (errmsg("postmaster was shut down, exiting")));
+		}
+
+		if (event->events & WL_SOCKET_CLOSED)
+		{
+			socketClosed = true;
+
+			/* we found what we are searching for */
+			break;
+		}
+
+		if (event->events & WL_LATCH_SET)
+		{
+			/*
+			 * A latch event might be preventing other events from being
+			 * reported.  Reset it and poll again.  No need to restore it
+			 * because no code should expect latches to survive across
+			 * CHECK_FOR_INTERRUPTS().
+			 */
+			ResetLatch(MyLatch);
+
+			goto retry;
```

Comment:
The way the code is structured doesn't align with the comment here.

Could you add a comment on why we are using `goto` to poll again?
Things that I find strange for example is that we do not reset the `eventIndex` back to `0` on retry.

Naively I would assume that if we 'poll again', we would need to iterate over all events that result from the new polling. However, that would only happen if the latch would be the first event reported by `WaitEventSetWait`, which might verywell _always_ be the case when it is hit, but still I would like to explicitly assign `eventIndex` its initial value again.

Also, if there is any way to refactor the `goto` away I would like that.

Something of the likes of
```C
bool socketClosed = false;
bool retry = false;
do {
  // poll events
  ...

  ProcessEvents(..., &socketClosed, &retry);
} while (retry);

return socketClosed;
```

Would be easier to reason about IMHO, `ProcessEvents` would iterate over all captured events and populate `retry` and `socketClosed` appropriately.

Reply:
> Things that I find strange for example is that we do not reset the eventIndex back to 0 on retry.

This is definitely a bug 

> Also, if there is any way to refactor the goto away I would like that.

Probably yes, my goal was to imitate the relevant PG function, but I guess on our code-base `goto` is much less common: https://github.com/postgres/postgres/blob/bcc8b14ef630b2ad9aae7813981fb248fbff9ed8/src/backend/libpq/pqcomm.c#L1921